File: python/00373_find_k_pairs_with_smallest_sums.py
class Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        if not nums1 or not nums2 or k == 0:
            return []

        heap = []
        result = []

        for i in range(min(k, len(nums1))):
            heapq.heappush(heap, (nums1[i] + nums2[0], nums1[i], nums2[0], 0))

        while k > 0 and heap:
            current = heapq.heappop(heap)
            num1 = current[1]
            num2 = current[2]
            index2 = current[3]

            result.append([num1, num2])

            if index2 + 1 < len(nums2):
                next_num2 = nums2[index2 + 1]
                heapq.heappush(heap, (num1 + next_num2, num1, next_num2, index2 + 1))

            k -= 1

        return result

File: java/00373_find_k_pairs_with_smallest_sums.java
class Solution {
    public List<List<Integer>> kSmallestPairs(int[] nums1, int[] nums2, int k) {
        List<List<Integer>> result = new ArrayList<>();
        if (nums1.length == 0 || nums2.length == 0 || k == 0) {
            return result;
        }

        PriorityQueue<int[]> minHeap = new PriorityQueue<>(
                (a, b) -> Integer.compare(a[0] + a[1], b[0] + b[1])
        );


        for (int i = 0; i < Math.min(k, nums1.length); i++) {
            minHeap.offer(new int[]{nums1[i], nums2[0], 0});
        }

        while (k-- > 0 && !minHeap.isEmpty()) {
            int[] curr = minHeap.poll();
            int num1 = curr[0];
            int num2 = curr[1];
            int index2 = curr[2];

            result.add(Arrays.asList(num1, num2));

            if (index2 + 1 < nums2.length) {
                minHeap.offer(new int[]{num1, nums2[index2 + 1], index2 + 1});
            }
        }

        return result;
    }

}

// Time complexity - O(k*logk)
// Space complexity - O (k)